File: src/game/GameEngine.ts
const TEMP_X5_DURATION_MS = 10 * 60 * 1000;

export interface RockLevel {
  level: number;
  name: string;
  tapsRequired: number;
  rewardCoins: number;
}

export type BoosterType = "TAP_X2" | "DYNAMITE_X100" | "TAP_X5_10_MIN";
export type ShopType = "BOOSTER" | "SKIN" | "BACKGROUND" | "SOUND";

export interface ShopItem {
  id: string;
  label: string;
  costCoins: number;
  shopType: ShopType;
  boosterType?: BoosterType;
}

export interface GameState {
  taps: number;
  coins: number;
  levelIndex: number;
  permanentMultiplier: number;
  tempX5UntilMs: number;
  equippedSkin: string;
  equippedBackground: string;
  equippedSound: string;
  ownedSkins: Set<string>;
  ownedBackgrounds: Set<string>;
  ownedSounds: Set<string>;
  achievements: Set<string>;
}

export const createInitialGameState = (): GameState => ({
  taps: 0,
  coins: 0,
  levelIndex: 0,
  permanentMultiplier: 1,
  tempX5UntilMs: 0,
  equippedSkin: "Default",
  equippedBackground: "Cantera",
  equippedSound: "Tap",
  ownedSkins: new Set(["Default"]),
  ownedBackgrounds: new Set(["Cantera"]),
  ownedSounds: new Set(["Tap"]),
  achievements: new Set()
});

export class GameEngine {
  constructor(private readonly now: () => number = () => Date.now()) {}

  readonly levels: RockLevel[] = [
    { level: 1, name: "Roca comun", tapsRequired: 0, rewardCoins: 0 },
    { level: 2, name: "Roca con carbon", tapsRequired: 1_000, rewardCoins: 50 },
    { level: 3, name: "Roca con hierro", tapsRequired: 10_000, rewardCoins: 250 },
    { level: 4, name: "Roca con cobre", tapsRequired: 25_000, rewardCoins: 600 },
    { level: 5, name: "Roca con oro", tapsRequired: 50_000, rewardCoins: 1_500 },
    { level: 6, name: "Roca con diamante", tapsRequired: 100_000, rewardCoins: 5_000 }
  ];

  readonly shopItems: ShopItem[] = [
    { id: "booster_x2", label: "Tap x2 permanente", costCoins: 150, shopType: "BOOSTER", boosterType: "TAP_X2" },
    {
      id: "booster_dyn",
      label: "Dinamita +100 taps",
      costCoins: 200,
      shopType: "BOOSTER",
      boosterType: "DYNAMITE_X100"
    },
    {
      id: "booster_x5",
      label: "Tap x5 por 10 min",
      costCoins: 500,
      shopType: "BOOSTER",
      boosterType: "TAP_X5_10_MIN"
    },
    { id: "skin_obsidian", label: "Skin Obsidiana", costCoins: 300, shopType: "SKIN" },
    { id: "skin_meteor", label: "Skin Meteorito", costCoins: 900, shopType: "SKIN" },
    { id: "bg_volcano", label: "Fondo Volcan", costCoins: 700, shopType: "BACKGROUND" },
    { id: "sound_metal", label: "Sonido metalico raro", costCoins: 250, shopType: "SOUND" }
  ];

  tap(state: GameState): GameState {
    const multiplier = state.tempX5UntilMs > this.now() ? 5 : 1;
    const addedTaps = state.permanentMultiplier * multiplier;
    return this.applyProgress({ ...state, taps: state.taps + addedTaps });
  }

  buyItem(state: GameState, itemId: string): GameState {
    const item = this.shopItems.find((candidate) => candidate.id === itemId);
    if (!item || state.coins < item.costCoins) {
      return state;
    }

    const paid = { ...state, coins: state.coins - item.costCoins };
    if (item.shopType === "BOOSTER") {
      if (!item.boosterType) {
        return state;
      }
      return this.applyBooster(paid, item.boosterType);
    }

    if (item.shopType === "SKIN") {
      return {
        ...paid,
        ownedSkins: new Set([...paid.ownedSkins, item.label]),
        equippedSkin: item.label
      };
    }

    if (item.shopType === "BACKGROUND") {
      return {
        ...paid,
        ownedBackgrounds: new Set([...paid.ownedBackgrounds, item.label]),
        equippedBackground: item.label
      };
    }

    return {
      ...paid,
      ownedSounds: new Set([...paid.ownedSounds, item.label]),
      equippedSound: item.label
    };
  }

  private applyBooster(state: GameState, boosterType: BoosterType): GameState {
    if (boosterType === "TAP_X2") {
      return { ...state, permanentMultiplier: state.permanentMultiplier * 2 };
    }

    if (boosterType === "DYNAMITE_X100") {
      return this.applyProgress({ ...state, taps: state.taps + 100 });
    }

    return {
      ...state,
      tempX5UntilMs: Math.max(state.tempX5UntilMs, this.now() + TEMP_X5_DURATION_MS)
    };
  }

  private applyProgress(state: GameState): GameState {
    const newLevelIndex = this.levels.reduce((acc, level, index) => {
      return state.taps >= level.tapsRequired ? index : acc;
    }, 0);

    const mergedAchievements = new Set([...state.achievements, ...this.computeAchievements(state.taps, newLevelIndex)]);
    if (newLevelIndex <= state.levelIndex) {
      return {
        ...state,
        achievements: mergedAchievements
      };
    }

    let reward = 0;
    for (let index = state.levelIndex + 1; index <= newLevelIndex; index += 1) {
      reward += this.levels[index].rewardCoins;
    }

    return {
      ...state,
      levelIndex: newLevelIndex,
      coins: state.coins + reward,
      achievements: mergedAchievements
    };
  }

  private computeAchievements(taps: number, levelIndex: number): Set<string> {
    const out = new Set<string>();
    if (levelIndex >= 1) out.add("Nivel 2: tu roca ahora parece util");
    if (levelIndex >= 2) out.add("Nivel 3: la roca ya paga impuestos");
    if (levelIndex >= 3) out.add("Nivel 4: desbloqueaste fuerza geologica");
    if (levelIndex >= 4) out.add("Nivel 5: tu roca ya revienta ventanas");
    if (levelIndex >= 5) out.add("Nivel 6: tu roca es leyenda mineral");
    if (taps >= 10_000) out.add("10k taps: dedos de titanio");
    if (taps >= 100_000) out.add("100k taps: tocar rocas es tu destino");
    return out;
  }
}

File: gamecore/src/main/kotlin/com/rockvolution/core/GameEngine.kt
package com.rockvolution.core

private const val TEMP_X5_DURATION_MS = 10 * 60 * 1000L

data class RockLevel(
    val level: Int,
    val name: String,
    val tapsRequired: Long,
    val rewardCoins: Int
)

enum class BoosterType { TAP_X2, DYNAMITE_X100, TAP_X5_10_MIN }

enum class ShopType { BOOSTER, SKIN, BACKGROUND, SOUND }

data class ShopItem(
    val id: String,
    val label: String,
    val costCoins: Int,
    val shopType: ShopType,
    val boosterType: BoosterType? = null
)

data class GameState(
    val taps: Long = 0,
    val coins: Int = 0,
    val levelIndex: Int = 0,
    val permanentMultiplier: Int = 1,
    val tempX5UntilMs: Long = 0,
    val equippedSkin: String = "Default",
    val equippedBackground: String = "Cantera",
    val equippedSound: String = "Tap",
    val ownedSkins: Set<String> = setOf("Default"),
    val ownedBackgrounds: Set<String> = setOf("Cantera"),
    val ownedSounds: Set<String> = setOf("Tap"),
    val achievements: Set<String> = emptySet()
)

class GameEngine(private val now: () -> Long = { System.currentTimeMillis() }) {
    val levels = listOf(
        RockLevel(1, "Roca común", 0, 0),
        RockLevel(2, "Roca con carbón", 1_000, 50),
        RockLevel(3, "Roca con hierro", 10_000, 250),
        RockLevel(4, "Roca con cobre", 25_000, 600),
        RockLevel(5, "Roca con oro", 50_000, 1_500),
        RockLevel(6, "Roca con diamante", 100_000, 5_000)
    )

    val shopItems = listOf(
        ShopItem("booster_x2", "Tap x2 permanente", 150, ShopType.BOOSTER, BoosterType.TAP_X2),
        ShopItem("booster_dyn", "Dinamita +100 taps", 200, ShopType.BOOSTER, BoosterType.DYNAMITE_X100),
        ShopItem("booster_x5", "Tap x5 por 10 min", 500, ShopType.BOOSTER, BoosterType.TAP_X5_10_MIN),
        ShopItem("skin_obsidian", "Skin Obsidiana", 300, ShopType.SKIN),
        ShopItem("skin_meteor", "Skin Meteorito", 900, ShopType.SKIN),
        ShopItem("bg_volcano", "Fondo Volcán", 700, ShopType.BACKGROUND),
        ShopItem("sound_metal", "Sonido metálico raro", 250, ShopType.SOUND)
    )

    fun tap(state: GameState): GameState {
        val multiplier = if (state.tempX5UntilMs > now()) 5 else 1
        val addedTaps = state.permanentMultiplier * multiplier
        return applyProgress(state.copy(taps = state.taps + addedTaps.toLong()))
    }

    fun buyItem(state: GameState, itemId: String): GameState {
        val item = shopItems.firstOrNull { it.id == itemId } ?: return state
        if (state.coins < item.costCoins) return state

        val paid = state.copy(coins = state.coins - item.costCoins)
        return when (item.shopType) {
            ShopType.BOOSTER -> applyBooster(paid, item.boosterType ?: return state)
            ShopType.SKIN -> paid.copy(ownedSkins = paid.ownedSkins + item.label, equippedSkin = item.label)
            ShopType.BACKGROUND -> paid.copy(
                ownedBackgrounds = paid.ownedBackgrounds + item.label,
                equippedBackground = item.label
            )

            ShopType.SOUND -> paid.copy(ownedSounds = paid.ownedSounds + item.label, equippedSound = item.label)
        }
    }

    fun equipSkin(state: GameState, skin: String): GameState =
        if (skin in state.ownedSkins) state.copy(equippedSkin = skin) else state

    fun equipBackground(state: GameState, background: String): GameState =
        if (background in state.ownedBackgrounds) state.copy(equippedBackground = background) else state

    fun equipSound(state: GameState, sound: String): GameState =
        if (sound in state.ownedSounds) state.copy(equippedSound = sound) else state

    private fun applyBooster(state: GameState, boosterType: BoosterType): GameState = when (boosterType) {
        BoosterType.TAP_X2 -> state.copy(permanentMultiplier = state.permanentMultiplier * 2)
        BoosterType.DYNAMITE_X100 -> applyProgress(state.copy(taps = state.taps + 100))
        BoosterType.TAP_X5_10_MIN -> state.copy(tempX5UntilMs = maxOf(state.tempX5UntilMs, now() + TEMP_X5_DURATION_MS))
    }

    private fun applyProgress(state: GameState): GameState {
        val newLevelIndex = levels.indexOfLast { state.taps >= it.tapsRequired }.coerceAtLeast(0)
        if (newLevelIndex <= state.levelIndex) {
            return state.copy(achievements = state.achievements + computeAchievements(state.taps, state.levelIndex))
        }

        var reward = 0
        for (index in (state.levelIndex + 1)..newLevelIndex) {
            reward += levels[index].rewardCoins
        }

        return state.copy(
            levelIndex = newLevelIndex,
            coins = state.coins + reward,
            achievements = state.achievements + computeAchievements(state.taps, newLevelIndex)
        )
    }

    private fun computeAchievements(taps: Long, levelIndex: Int): Set<String> {
        val out = mutableSetOf<String>()
        if (levelIndex >= 1) out += "Nivel 2: tu roca ahora parece útil"
        if (levelIndex >= 2) out += "Nivel 3: la roca ya paga impuestos"
        if (levelIndex >= 3) out += "Nivel 4: desbloqueaste fuerza geológica"
        if (levelIndex >= 4) out += "Nivel 5: tu roca ya revienta ventanas"
        if (levelIndex >= 5) out += "Nivel 6: tu roca es leyenda mineral"
        if (taps >= 10_000) out += "10k taps: dedos de titanio"
        if (taps >= 100_000) out += "100k taps: tocar rocas es tu destino"
        return out
    }
}
